[EDU-2305] feat: add object storage API reference (#545)
aziontech/docs: src/content/docs/en/pages/devtools/azion-edge-runtime/api-reference/object-storage/object-storage.mdx

---
title: Object Storage API 
description: The Edge Functions Storage API is an interface that allows access to the Azion Object Storage. Through this interface, it's possible to read and write data in the storage and its buckets.
meta_tags: edge function, edge computing, storage, jamstack, apis
namespace: documentation_runtime_storage
permalink: /documentation/devtools/runtime/api-reference/storage/
menu_namespace: runtimeMenu
---

The [Edge Functions](/en/documentation/products/edge-application/edge-functions/) Storage API is an interface that allows access to the Azion Object Storage. Through this interface, it's possible to read and write data in the storage and its buckets.

## Class instantiation

To leverage this API, you need to import and instantiate the `Storage` class, passing the bucket's name.

### Syntax

Import: 

```javascript
import Storage from "azion:storage";
```

Instantiate: 

```js
const storage = new Storage(bucket);
```

### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `bucket` | string | Name of a bucket created during an [Azion CLI](/en/documentation/products/azion-cli/overview/) deploy or through [Real-Time Manager (RTM)](/en/documentation/products/guides/getting-to-know-real-time-manager/). |

Reviewer: ```suggestion
| `bucket` | string | Name of a bucket created during an [Azion CLI](/en/documentation/products/azion-cli/overview/) deploy or through [Real-Time Manager (RTM)](/en/documentation/products/guides/getting-to-know-real-time-manager/) |
```

        return new Response(error, {status:500});
    }
    return new Response("Ok");
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

---

### async Storage.list()

The `Storage.list()` method is used to list all objects belonging to the specific `bucket`.

:::note
This is an async method and must be used with await or event.waitUntil.
:::

#### Syntax 

```js
await storage.list()
```

#### Return 

| Success | Error | 
|-|-|
| Returns an object of the class StorageObjectList | It throws a `StorageError`|

#### Code sample

1. Listing objects from Azion Storage:

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const objectsList = await storage.list();
        for (const entry of objectsList.entries) {
            console.log(`key: ${entry.key} length: ${entry.content_length}`);
        }
        return new Response("Ok");   
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});

> ---

Reviewer: verificar que o bloco de código esteja fechando aqui

### StorageObject Class Methods

| Method | Description |
|-|-|
| `StorageObject.content` | Read-only property containing a ReadableStream for the content stored in the storage |
| `async StorageObject.arrayBuffer()` | Asynchronous function that returns an ArrayBuffer with the content stored in the storage. This method consumes the ReadableStream from the content property |

> | `StorageObject.metadata` | Read-only property containing a [Map object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) with the metadata of the object stored in the storage |

Reviewer: o link não está funcionando para mim

Author: tenta puxar a branch local e rodar... a principio ta funcionando bem

---
title: Object Storage API 
description: The Edge Functions Storage API is an interface that allows access to the Azion Object Storage. Through this interface, it's possible to read and write data in the storage and its buckets.
meta_tags: edge function, edge computing, storage, jamstack, apis
namespace: documentation_runtime_storage
permalink: /documentation/devtools/runtime/api-reference/storage/
menu_namespace: runtimeMenu
---

> The [Edge Functions](/en/documentation/products/edge-application/edge-functions/) storage API is an interface that allows access to the Azion Object Storage. Through this interface, it's possible to read and write data in the storage and its buckets.

Reviewer: ```suggestion
The [Edge Functions](/en/documentation/products/edge-application/edge-functions/) Storage API is an interface that allows access to the Azion Object Storage. Through this interface, it's possible to read and write data in the storage and its buckets.
```

## Class instantiation

> 

Reviewer: ```suggestion
```

To leverage this API, you need to import and instantiate the `Storage` class, passing the bucket's name.

### Syntax

Import: 

```javascript
import Storage from "azion:storage";
```

Instantiate: 

```js
const storage = new Storage(bucket);
```

### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `bucket` | string | Name of a bucket created during an [Azion CLI](/en/documentation/products/azion-cli/overview/) deploy or through [RTM](/en/documentation/products/guides/getting-to-know-real-time-manager/). |

Reviewer: ```suggestion
| `bucket` | string | Name of a bucket created during an [Azion CLI](/en/documentation/products/azion-cli/overview/) deploy or through [Real-Time Manager (RTM)](/en/documentation/products/guides/getting-to-know-real-time-manager/). |
```

Reviewer: por ser a primeira vez que aparece o termo

### Return value

Storage object used to access Azion Object Storage.

---

## Methods

### async Storage.put

The `Storage.put` method is used to insert a new object into the storage.

:::note

> This is an async method, and must be used with `await` or `event.waitUntil`.

Reviewer: ```suggestion
This is an async method and must be used with `await` or `event.waitUntil`.
```

:::

#### Syntax

```js
async Storage.put(key, value, options)
```

#### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `key` | string | Identifier that allows the search for an object in the storage. |

Reviewer: ```suggestion
| `key` | string | Identifier that allows the search for an object in the storage |
```

> | `value` | ArrayBuffer ou ReadableStream | Content of the object being stored. In case `value` implements a [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream), the Stream APIs can be used and the option `content-lenght` is required. |

Reviewer: ```suggestion
| `value` | ArrayBuffer ou ReadableStream | Content of the object being stored. In case `value` implements a [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream), the Stream APIs can be used and the option `content-lenght` is required |
```

> | `options` | object | The `options` attributes are described in the table below. |

Reviewer: ```suggestion
| `options` | object | The `options` attributes are described in the table below |
```

**options**

To pass the `options` parameter, you must provide an object with the following attributes:

| Attribute | Type | Description |
| - | - | - |

> | `content-type` | string | The `content-type` of the object being created. It's similar to the HTTP Header `content-type` and describes how the content of the object storage should be treated. `content-type` is optional; the default value is `application/octet-stream`. |

Reviewer: ```suggestion
| `content-type` | string | The `content-type` of the object being created. It's similar to the HTTP Header `content-type` and describes how the content of the object storage should be treated. `content-type` is optional; the default value is `application/octet-stream` |
```

> | `content-length` | string | The size of the object to be created, in bytes. It's mandatory when the value is ReadableStream. |

Reviewer: ```suggestion
| `content-length` | string | The size of the object to be created, in bytes. It's mandatory when the value is `ReadableStream` |
```

| `metadata` | object | Any JavaScript object containing information about the object that will be stored in the Storage. All properties of the object must be strings. |

#### Return

| Success | Error |
|-|-|
| No response | Throws a `StorageError`|

#### Code sample

> 1. To persist the request body:

Reviewer 2: será que poderia ser `-` ao invés de números? só pra não dar a ideia de que serão consequência dos anteriores

});

```

3. To save a JSON with metadata:

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test2";
        const data = JSON.stringify({
            name:"Paul",
            address:"Abbey Road"
        });
        const buffer = new TextEncoder().encode(data);
        const options = { "content-type": "application/json",
                          "metadata": {
                            info:"test info",
                            counter: "1",
                          }}
        await storage.put(key, buffer, options);
        return new Response("OK");
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

---

### async storage.get(key)

The `storage.get(key)` method is used to retrieve an object.

:::note
The method is async and must be used with `await` or `event.waitUntil`.
:::

#### Syntax 

```js
async storage.get(key)
```

#### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `key` | string | Identifier that allows the search for an object in the bucket. |

Reviewer: ```suggestion
| `key` | string | Identifier that allows the search for an object in the bucket |
```

#### Return 

| Success | Error | 
|-|-|
| Returns an object of the class `StorageObject` | Throws a `StorageError`|

> 

Reviewer 2: ```suggestion
```

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test";
        const storageObject = await storage.get(key); 
        const data = new TextDecoder().decode(await storageObject.arrayBuffer());
        return new Response(data);
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

3. To read an object's metadata:

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test2"
        const storageObject = await storage.get(key);
        return new Response(storageObject.metadata.get("info"));   
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
``` 

---

### async Storage.delete(key)

The `Storage.delete(key)` is used to remove an object from Azion Object Storage.

:::note

> The method is async and must be used with `await` or `event.waitUntil`.

Reviewer 2: ```suggestion
This is an async method and must be used with `await` or `event.waitUntil`.
```

:::

#### Syntax 

```js
async Storage.delete(key)
```

#### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `key` | string | Identifier that allows the search for an object in the bucket. |

Reviewer: ```suggestion
| `key` | string | Identifier that allows the search for an object in the bucket |
```

#### Return 

| Success | Error | 
|-|-|
| No response | Throws a `StorageError`|

#### Code sample

1. To remove an object stored in Azion Object Storage related to a specific `bucket`:

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    const bucket = "mybucket";
    const storage = new Storage(bucket);
    const key = "test";
    try{
        await storage.delete(key);
    }catch(error){
        return new Response(error, {status:500});
    }
    return new Response("Ok");
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

---

### async Storage.list()

> the `Storage.list()` method is used to list all objects belonging to the specific `bucket`.

Reviewer: ```suggestion
The `Storage.list()` method is used to list all objects belonging to the specific `bucket`.
```

:::note

> The method is async, and must be used with await or event.waitUntil.

Reviewer: ```suggestion
The method is async and must be used with `await` or `event.waitUntil`.
```

> The method is async, and must be used with await or event.waitUntil.

Reviewer 2: ```suggestion
This is an async method and must be used with await or event.waitUntil.
```

:::

#### Syntax 

```js
await storage.list()
```

#### Return 

| Success | Error | 
|-|-|
| Returns an object of the class StorageObjectList | It throws a `StorageError`|

#### Code sample

1. Listing objects from Azion Storage:

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const objectsList = await storage.list();
        for (const entry of objectsList.entries) {
            console.log(`key: ${entry.key} length: ${entry.content_length}`);
        }
        return new Response("Ok");   
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});

> ```

Reviewer 2: ```suggestion
```

```

---

### StorageObject Class Methods

> |Method|Description|

Reviewer 2: ```suggestion
| Method | Description |
```

|-|-|

> |`StorageObject.content`|Read-only property containing a ReadableStream for the content stored in the storage.|

Reviewer: ```suggestion
| `StorageObject.content` | Read-only property containing a ReadableStream for the content stored in the storage |
```

|`async StorageObject.arrayBuffer()`| Asynchronous function that returns an ArrayBuffer with the content stored in the storage. This method consumes the ReadableStream from the content property.|

> |`StorageObject.metadata`| Read-only property containing a [Map object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) with the metadata of the object stored in the storage.|

Reviewer: ```suggestion
| `StorageObject.metadata` | Read-only property containing a [Map object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) with the metadata of the object stored in the storage |
```

> | `StorageObject.contentType`|Read-only property containing the `content-type` of the content stored in the storage.|

Reviewer: ```suggestion
| `StorageObject.contentType` | Read-only property containing the `content-type` of the content stored in the storage |
```

> | `StorageObject.contentLength`|  Read-only property containing the size in bytes of the content stored in the storage.|

Reviewer: ```suggestion
| `StorageObject.contentLength` |  Read-only property containing the size in bytes of the content stored in the storage |
```

---
title: Object Storage API 

> description: The Edge Functions Storage API is an interface that allows access to Azion Storage. Through this interface, it's possible to write and retrieve data in the Storage.

Reviewer 3: ```suggestion
description: The Edge Functions Storage API is an interface that allows access to the Azion Object Storage. Through this interface, it's possible to write and retrieve data in the bucket.
```

Reviewer 3: Tem que confirmar o nome oficial, mas acho que é Object Storage mesmo.

meta_tags: edge function, edge computing, storage, jamstack, apis
namespace: documentation_runtime_storage
permalink: /documentation/devtools/runtime/api-reference/storage/
menu_namespace: runtimeMenu
---

> The [Edge Functions](/en/documentation/products/edge-application/edge-functions/) storage API is an interface that allows access to Azion Storage. Through this interface, it's possible to write and retrieve data.

Reviewer 3: ```suggestion
The [Edge Functions](/en/documentation/products/edge-application/edge-functions/) storage API is an interface that allows access to the Azion Object Storage. Through this interface, it's possible to read and write data in the bucket.
```

## Class instantiation

To leverage this API, you need to import and instantiate the Storage class, passing the bucket's name.

### Syntax

Import: 

```javascript
import Storage from "azion:storage";
```

Instantiate: 

```js
const storage = new Storage(bucket);
```

### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `bucket` | string | Name of a bucket created during an [Azion CLI](/en/documentation/products/azion-cli/overview/) deploy or through [RTM](/en/documentation/products/guides/getting-to-know-real-time-manager/). |

Reviewer 3: Acho que por enquanto vai ser só API/CLI, tem que conferir com o Vitor se tá planejado ter interface.

### Return value

> Storage object used to access Azion Storage

Reviewer 3: ```suggestion
Storage object used to access Azion Storage.
```

---

## Methods

### async Storage.put

> The Storage.put method is used to insert a new object to Azion Storage.

Reviewer 3: ```suggestion
The `Storage.put` method is used to insert a new object to Object Storage.
```

:::note

> This is an async method, and must be used with await or event.waitUntil.

Reviewer 3: ```suggestion
This is an async method, and must be used with `await` or `event.waitUntil`.
```

:::

#### Syntax

```js
async Storage.put(key, value, options)
```

#### Parameters

| Parameter | Type | Description |
| - | - | - |
| `key` | string | Identifier that allows the search for an object in the Storage. |

> | `value` | ArrayBuffer ou ReadableStream | Content of the object being stored. In case `value` implements a [ReadableStream](), the Stream APIs can be used and the option `content-lenght` is required. |

Author: ```suggestion
| `value` | ArrayBuffer ou ReadableStream | Content of the object being stored. In case `value` implements a [ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream), the Stream APIs can be used and the option `content-lenght` is required. |
```

| `options` | object | The `options` attributes are described in the table below. |

**options**

To pass the `options` parameter, it's necessary to provide an object with the following attributes:

| Attribute | Type | Description |
| - | - | - |
| `content-type` | string | The `content-type` of the object being created. It's similar to the HTTP Header `content-type` and describes how the content of the object storage should be treated. `content-type` is optional; the default value is `application/octet-stream`. |
| `content-length` | string | The size of the object to be created, in bytes. It's mandatory when the value is ReadableStream. |
| `metadata` | object | Any JavaScript object containing information about the object that will be stored in the Storage. All properties of the object must be strings. |

#### Return

| Success | Error |
|-|-|
| No response| It throws a `StorageError`|

#### Code sample

> 1. Persisting the request body:

Reviewer 3: ```suggestion
1. To persist the request body:
```

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = “mybucket” 
        const storage = new Storage(bucket);
        const key = "test";
        const inputStream = event.request.body;
        let contentLength = event.request.headers.get("content-length");
        await storage.put(key, inputStream, { "content-length": contentLength });
        return new Response("OK");
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

> 2. Persisting a JSON object:

Reviewer 3: ```suggestion
2. To persist a JSON object:
```

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test";
        const data = JSON.stringify({
            name:"John",

>             address:"Abbey Road"

Reviewer 3: Que exemplo fofo hehehe

        });
        const buffer = new TextEncoder().encode(data);
        await storage.put(key, buffer);
        return new Response("OK");
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});

```

> 3. Saving a JSON with metadata:

Reviewer 3: ```suggestion
3. To save a JSON with metadata:
```

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test2";
        const data = JSON.stringify({
            name:"Paul",
            address:"Abbey Road"
        });
        const buffer = new TextEncoder().encode(data);
        const options = { "content-type": "application/json",
                          "metadata": {
                            info:"test info",
                            counter: "1",
                          }}
        await storage.put(key, buffer, options);
        return new Response("OK");
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

---

### async storage.get(key)

the `storage.get(key)` method is used to retrieve an object.

:::note
The method is async, and must be used with await or event.waitUntil.
:::

#### Syntax 

```js
async storage.get(key)
```

#### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `key` | string | Identifier that allows the search for an object in the Storage. |

Reviewer 3: ```suggestion
| `key` | string | Identifier that allows the search for an object in the bucket. |
```

#### Return 

| Success | Error | 
|-|-|

> | Returns an object of the class `StorageObject` | It throws a `StorageError`|

Reviewer 3: ```suggestion
| Returns an object of the class `StorageObject` | Throws a `StorageError`|
```

#### Code sample

> 1. Retrieving content from the storage and returning it:

Reviewer 3: ```suggestion
1. To retrieve and return content from the storage:
```

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test";
        const storageObject = await storage.get(key);
        return new Response(storageObject.content);
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

> 2. Getting a JSON from the storage:

Reviewer 3: ```suggestion
2. To get a JSON from the storage:
```

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test";
        const storageObject = await storage.get(key); 
        const data = new TextDecoder().decode(await storageObject.arrayBuffer());
        return new Response(data);
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```

> 3. Reading an object's metadata:

Reviewer 3: ```suggestion
3. To read an object's metadata:
```

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const key = "test2"
        const storageObject = await storage.get(key);
        return new Response(storageObject.metadata.get("info"));   
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
``` 

---

### async Storage.delete(key)

> the `storage.get(key)` method is used to retrieve an object.

Reviewer 3: ```suggestion
The `storage.get(key)` method is used to retrieve an object.
```

:::note
The method is async, and must be used with await or event.waitUntil.
:::

#### Syntax 

> The `Storage.delete(key)` is used to remove an object from Azion Storage.

Reviewer 3: ```suggestion
The `Storage.delete(key)` is used to remove an object from Azion Object Storage.
```

```js
async Storage.delete(key)
```

#### Parameters

| Parameter | Type | Description |
| - | - | - |

> | `key` | string | Identifier that allows the search for an object in the Storage. |

Reviewer 3: ```suggestion
| `key` | string | Identifier that allows the search for an object in the bucket. |
```

#### Return 

| Success | Error | 
|-|-|

> | Returns an object of the class `StorageObjectList` | It throws a `StorageError`|

Reviewer 3: ```suggestion
| Returns an object of the class `StorageObjectList` | Throws a `StorageError`|
```

#### Code sample

> 1. Listing objects stored in Azion Storage related to a specific `bucket`:

Reviewer 3: ```suggestion
1. To list objects stored in Azion Object Storage related to a specific `bucket`:
```

:::note
The method is async, and must be used with await or event.waitUntil.
:::

#### Syntax 

The `Storage.delete(key)` is used to remove an object from Azion Storage.
```js
await storage.list()
```

#### Parameters

| Parameter | Type | Description |
| - | - | - |
| `key` | string | Identifier that allows the search for an object in the Storage. |

#### Return 

| Success | Error | 
|-|-|
| No response | It throws a `StorageError`|

#### Code sample

1. Removing an object from Azion Storage:

```js
import Storage from "azion:storage";

async function handleRequest(event) {
    try{
        const bucket = "mybucket";
        const storage = new Storage(bucket);
        const objectsList = await storage.list();
        for (const entry of objectsList.entries) {
            console.log(`key: ${entry.key} length: ${entry.content_length}`);
        }
        return new Response("Ok");   
    }catch(error){
        return new Response(error, {status:500});
    }
}

addEventListener("fetch", (event) => {
    event.respondWith(handleRequest(event));
});
```
---

### StorageObject Class Methods

|Method|Description|
|-|-|

> |`StorageObject.content`|Read-only property containing a ReadableStream for the content stored in the storage.|
> |`async StorageObject.arrayBuffer()`| Asynchronous function that returns an ArrayBuffer with the content stored in the storage. This method consumes the ReadableStream from the content property.|

Reviewer 3: ```suggestion
|`StorageObject.content`|Read-only property containing a `ReadableStream` for the content stored in the storage.|
|`async StorageObject.arrayBuffer()`| Asynchronous function that returns an `ArrayBuffer` with the content stored in the storage. This method consumes the `ReadableStream` from the content property.|
```
